### PR TITLE
feat: support JSON data type

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for JSON data type.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcDataType.java
Comment (olavloite) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcDataType.java:235]: JDBC does not have anything like a JSON type, and we need to have a specific type code for this in order to be able to set JSON values as parameters on `PreparedStatement`. Note that this is also what is needed when working with for example PostgreSQL through JDBC: https://stackoverflow.com/questions/35844138/how-can-i-insert-json-object-into-postgres-using-java-preparedstatement

File: src/main/resources/com/google/cloud/spanner/jdbc/DatabaseMetaData_GetColumns.sql
Comment (olavloite) [inline src/main/resources/com/google/cloud/spanner/jdbc/DatabaseMetaData_GetColumns.sql:27]: According to the JDBC specification this should always return one of the constants in `java.sql.Types`. That means that we should not return the vendor code of the JSON type here.

File: src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcDatabaseMetaDataTest.java
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcDatabaseMetaDataTest.java:125]: `2621440` is equal to the length of a `STRING(MAX)` column. If JSON has a different max length, then this should be changed.

**Comment by olavloite** [comment]
@zoercai @thiagotnunes This is an initial implementation for supporting JSON in JDBC based on the current state of https://github.com/googleapis/java-spanner/pull/872. It assumes the following:
1. JSON will be treated as a String whenever possible. So for example to get a JSON value from a `ResultSet`, you would normally call `ResultSet#getString(int)` (or alternatively `ResultSet#getObject(int)` which will also return the JSON value as a String).
2. When setting a JSON value as a parameter on a `PreparedStatement`, we need to make a distinction between String and JSON. That is because the backend does not support automatic coercion from String to JSON, which means that the JDBC driver somehow needs to know that the String value that was set should in some way be treated as JSON. That can be achieved using one of the following strategies:
2.1. Call `PreparedStatement#setObject(int, Object, int)` where the value is passed as a String and the last parameter is the vendor type code for JSON in Cloud Spanner (`com.google.cloud.spanner.jdbc.JsonType.VENDOR_TYPE_NUMBER`).
2.2. Call `PreparedStatement#setObject(int, Object)` where the value is passed as a `Value.json("some-json-string")`.

The above handling is [similar to how this is also handled](https://stackoverflow.com/questions/35844138/how-can-i-insert-json-object-into-postgres-using-java-preparedstatement) in the JDBC driver of PostgreSQL.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java
Comment (thiagotnunes) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:207]: That is an interesting design decision. So if we do `preparedStatement.setObject(Value.json("{...}"))` we would need to specify the `SqlType`. I like this.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java
Comment (thiagotnunes) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: Why do we use `ASCII` instead of `UTF-8` here?

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java
Comment (thiagotnunes) [inline src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java:591]: I think this is the way to go, but more of a design question: do we ever want to externalise `Value.json` as a return type as well?

**Review by thiagotnunes — APPROVED** [review]

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java
Comment (thiagotnunes) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: cc @zoercai

Comment (zoercai) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: Thanks @thiagotnunes , yep String values passed to the backend must be UTF-8 encoded

Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: This method is used for this (required JDBC interface) method: https://github.com/googleapis/java-spanner-jdbc/blob/4ed17f0061077505d772c89589a0965f3bbc4a18/src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java#L361

There's also a similar Unicode version which is called [setCharacterStream(..)](https://github.com/googleapis/java-spanner-jdbc/blob/4ed17f0061077505d772c89589a0965f3bbc4a18/src/main/java/com/google/cloud/spanner/jdbc/AbstractJdbcPreparedStatement.java#L373), which expects the input to be in Unicode format.

The data is __read__ as ASCII, but encoded as UTF-8 when it is sent to Cloud Spanner.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java
Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java:591]: Good point.

What we could do is modify the implementation of this method: https://github.com/googleapis/java-spanner-jdbc/blob/4ed17f0061077505d772c89589a0965f3bbc4a18/src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java#L896

It currently supports converting between common Java types, e.g. converting a `Timestamp` to a `String`. We could add support for specifying `com.google.cloud.spanner.Value.class` as the type, and then return the value in the result set as a `Value`.

Comment (thiagotnunes) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java:591]: That makes sense to me. Does not necessary need to be a change added to this PR if you feel it should be out of scope here.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java
Comment (thiagotnunes) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: Ah ok, this method is only used in the `setAsciiStream` then? I guess that should be fine if that is required by JDBC.

Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcParameterStore.java:675]: > Ah ok, this method is only used in the `setAsciiStream` then? I guess that should be fine if that is required by JDBC.

Yes, this is only used by the `setAsciiStream` method. JDBC unfortunately contains a number of old method definitions that still need to be supported.

File: src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java
Comment (olavloite) [reply src/main/java/com/google/cloud/spanner/jdbc/JdbcResultSet.java:591]: I think that should be a separate change. I'll create a PR specifically for that, and once that is merged, add it for JSON here as well.

**Comment by zoercai** [comment]
Hi @olavloite , I've merged the JSON Java PR and started the release https://github.com/googleapis/java-spanner/pull/1364. Please rebase and rerun CI when the release is done, and merge this PR once CI passes. Thanks!

**Comment by zoercai** [comment]
update: the release is done.

**Pull request merged**
